examples/nlp/duplex_text_normalization/analyze_errors.py

```diff
@@ -0,0 +1,233 @@
+        for idx in range(nb_tokens):
+            if idx == nb_tokens - 1:
+                if tokens_hightlight[idx] != cur_bool_val:
+                    spans.append((cur_start_idx, nb_tokens - 2, cur_bool_val))
+                    spans.append((nb_tokens - 1, nb_tokens - 1, tokens_hightlight[idx]))
+                else:
+                    spans.append((cur_start_idx, nb_tokens - 1, cur_bool_val))
+            else:
+                if tokens_hightlight[idx] != cur_bool_val:
+                    spans.append((cur_start_idx, idx - 1, cur_bool_val))
+                    cur_start_idx, cur_bool_val = idx, tokens_hightlight[idx]
+        return spans
+
+    def get_spans_html(self, spans, tokens):
+        html_str = ''
+        for start, end, type in spans:
+            color = 'red' if type else 'black'
+            span_tokens = tokens[start : end + 1]
+            span_str = '<span style="color:{}">{}</span> '.format(color, ' '.join(span_tokens))
+            html_str += span_str
+        return html_str
+
+
+# Main function for analysis
+def analyze(errors_log_fp, visualization_fp):
+    # Read lines from errors log
+    with open(errors_log_fp, 'r', encoding='utf-8') as f:
+        lines = f.readlines()
+
+    # Process lines
+    tn_error_cases, itn_error_cases = [], []
+    for ix in range(0, len(lines), 7):
+        mode_line = lines[ix]
+        info_lines = lines[ix + 1 : ix + 6]
+        # Append new error case
+        if mode_line.startswith('Forward Problem'):
+            mode = constants.TN_MODE
+            tn_error_cases.append(ErrorCase.from_lines(info_lines, mode))
+        elif mode_line.startswith('Backward Problem'):
+            mode = constants.ITN_MODE
+            itn_error_cases.append(ErrorCase.from_lines(info_lines, mode))
+
+    # Basic stats
+    print('---- Text Normalization ----')
+    print('Number of TN errors: {}'.format(len(tn_error_cases)))
+
+    print('---- Inverse Text Normalization ---- ')
+    print('Number of ITN errors: {}'.format(len(itn_error_cases)))
+
+    # Produce a visualization
+    with open(visualization_fp, 'w+', encoding='utf-8') as f:
+        # Appendix
+        f.write('Appendix</br>')
+        f.write('<a href="#tn_section">Text Normalization Analysis.</a></br>')
+        f.write('<a href="#itn_section">Inverse Text Normalization Analysis.</a>')
+
+        # TN Section
+        f.write('<h2 id="tn_section">Text Normalization</h2>\n')
+        for errorcase in tn_error_cases:
+            f.write(errorcase.get_html())
+
+        # ITN Section
+        f.write('<h2 id="itn_section">Inverse Text Normalization</h2>\n')
+        for errorcase in itn_error_cases:
+            f.write(errorcase.get_html())
+
+
+if __name__ == '__main__':
+    # Parse argument
+    parser = ArgumentParser()
+    parser.add_argument('--errors_log_fp', required=True)
+    parser.add_argument('--visualization_fp', required=True)
+    args = parser.parse_args()
```

Comment:
add help and types to the args + docstrings across the file

Reply:
I added comments and docstrings.